common/python/ax/cloud/aws/aws_s3.py

```diff
@@ -106,27 +119,37 @@ def get_bucket_name(self):
         stop_max_attempt_number=3
     )
     def _get_bucket_region_from_aws(self):
-        # We assume cluster is not access any resource outside partition, e.g.
-        # clusters in partition "aws" will not access resource in partition "aws-us-gov"
-        instance_region = Cloud().meta_data().get_region()
-        s3 = boto3.Session(
-            profile_name=self._aws_profile,
-            region_name=instance_region
-        ).client("s3", config=Config(signature_version='s3v4'))
-
-        logger.debug("Finding region for bucket %s from with initial region %s", self._name, instance_region)
-        try:
-            response = s3.head_bucket(Bucket=self._name)
-            logger.debug("Head_bucket returned OK %s", response)
-        except ClientError as e:
-            if "Not Found" in str(e):
+
+        def _do_get_region(start_region):
+            s3 = boto3.Session(
+                profile_name=self._aws_profile,
+                region_name=start_region
+            ).client("s3", config=Config(signature_version='s3v4'))
+
+            logger.debug("Finding region for bucket %s from with initial region %s", self._name, start_region)
+            try:
+                response = s3.head_bucket(Bucket=self._name)
+                logger.debug("Head_bucket returned OK %s", response)
+            except ClientError as e:
+                response = getattr(e, "response", {})
+                logger.debug("Head_bucket returned error %s, inspecting headers", response)
                 return None
-            response = getattr(e, "response", {})
-            logger.debug("Head_bucket returned error %s, inspecting headers", response)
-        headers = response.get("ResponseMetadata", {}).get("HTTPHeaders", {})
-        region = headers.get("x-amz-bucket-region", headers.get("x-amz-region", None))
-        logger.debug("Found region %s from head_bucket for %s, headers %s", region, self._name, headers)
-        return region
+            headers = response.get("ResponseMetadata", {}).get("HTTPHeaders", {})
+            region = headers.get("x-amz-bucket-region", headers.get("x-amz-region", None))
+            logger.debug("Found region %s from head_bucket for %s, headers %s", region, self._name, headers)
+            return region
+
+        if Cloud().own_cloud() == Cloud.CLOUD_AWS:
+            # When running on AWS instance, we query metadata server for initial region to get bucket region
+            return _do_get_region(Cloud().meta_data().get_region())
+        else:
+            # Assume we don't have AWS metadata server access
+            for r in PARTITION_DEFAULT_REGIONS:
+                logger.debug("Trying partition default region %s to get bucket region.", r)
+                bucket_region = _do_get_region(r)
```

Comment:
Is this still not a singleton? If that is case, this is quite heavy weight operation. Is there any performance impact?

Reply:
This code path will only get touched when operating a cluster from outside the cluster, i.e. non-aws environment. Currently this is the only way I can think of to guess partition without asking additional inputs.

For cluster in aws partition (which is the normal case) will only call `_do_get_region` once; aws-us-gov partition (which is very rare), will cal it twice, aws-cn partition is a placeholder as we haven't seen any installation there.

I agree that making this class a singleton (regarding bucket name) will improve performance but is out of scope of this PR, and this purpose of this PR is to make things correct